freqtrade/rpc/emc.py

```diff
@@ -0,0 +1,234 @@
+                                    logger.info(
+                                        f"Ping error {url} - retrying in {self.sleep_time}s")
+                                    await asyncio.sleep(self.sleep_time)
+
+                                    break
+                            except Exception as e:
+                                logger.exception(e)
+                                continue
+                except (
+                    socket.gaierror,
+                    ConnectionRefusedError,
+                    websockets.exceptions.InvalidStatusCode
+                ) as e:
+                    logger.error(f"Connection Refused - {e} retrying in {self.sleep_time}s")
+                    await asyncio.sleep(self.sleep_time)
+
+                    continue
+
+        except asyncio.CancelledError:
+            # Exit silently
+            pass
+
+    def compose_consumer_request(self, type_: str, data: Any) -> Dict[str, Any]:
+        """
+        Create a request for sending to a producer
+
+        :param type_: The RPCRequestType
+        :param data: The data to send
+        :returns: Dict[str, Any]
+        """
+        return {'type': type_, 'data': data}
+
+    # How we do things here isn't set in stone. There seems to be some interest
+    # in figuring out a better way, but we shall do this for now.
+    def handle_producer_message(self, message: Dict[str, Any]):
+        """
+        Handles external messages from a Producer
+        """
+        # Should we have a default message type?
+        message_type = message.get('type', RPCMessageType.STATUS)
+        message_data = message.get('data')
+
+        # We shouldn't get empty messages
+        if message_data is None:
+            return
+
+        logger.debug(f"Received message of type {message_type}")
+
+        # Handle Whitelists
+        if message_type == RPCMessageType.WHITELIST:
+            pairlist = message_data
+
+            # Add the pairlist data to the ExternalPairlist plugin
+            external_pairlist = self._rpc._freqtrade.pairlists._pairlist_handlers[0]
+            external_pairlist.add_pairlist_data(pairlist)
+
+        # Handle analyzed dataframes
+        elif message_type == RPCMessageType.ANALYZED_DF:
+            key, value = message_data.get('key'), message_data.get('value')
+
+            if key and value:
+                pair, timeframe, candle_type = key
+
+                # Convert the JSON to a pandas DataFrame
+                dataframe = json_to_dataframe(value)
+
+                # If set, remove the Entry and Exit signals from the Producer
+                if self._emc_config.get('remove_entry_exit_signals', False):
+                    dataframe = remove_entry_exit_signals(dataframe)
+
+                # Add the dataframe to the dataprovider
+                dataprovider = self._rpc._freqtrade.dataprovider
+                dataprovider.add_external_df(pair, timeframe, dataframe, candle_type)
```

Comment:
you will want to attach dataprovider to this class directly (shouldn't be a problem when done on init).

The way you're taking here is clearly not acceptable - as it's going through 3 classes to get the class you actually need.

Reply:
:+1: